registry/client/transport/http_reader.go

```diff
@@ -2,15 +2,31 @@ package transport
 
 import (
 	"bufio"
-	"bytes"
 	"errors"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"net/http"
 	"os"
 )
 
+// HTTPError encapsulates an unsuccessful HTTP status code.
+type HTTPError struct {
+	Resp       *http.Response
+	bodyClosed bool
+}
+
+// Error renders an HTTPError to a string.
+func (h *HTTPError) Error() string {
+	// If the Error method gets called, the error propagated somewhere
+	// where it's not being handled natively as a HTTPError, so we need
+	// to take care of closing the body.
+	if !h.bodyClosed {
+		h.Resp.Body.Close()
```

Comment:
I guess the question here is whether it's reasonable to close the body on the next `Read` call. For example, should we allow code that does this:

```go
for ... {
        _, err := readSeekCloser.Read(p)
         if err != nil {
                if httpError, ok := err.(*transport.HTTPError); ok {
                        go doSomethingAsyncWithTheBody(httpError.Resp.Body)
                }
        }
}
```

This would be unsafe in the model where we only allow one open response body a time. But it seems like a weird case, and thinking about it some more, breaking this kind of code is probably better than potentially keeping many sockets open until the `HTTPReadSeekCloser` is closed, potentially hitting resource limits in the mean time.

So I'll move the body closing from `Error` to `Close`/`reader` and keep at most one response body open at a time, if you think that's reasonable.